### PR TITLE
WEBAPP-553: Build configs

### PR DESCRIPTION
**Pull request by klinzo**
This pull request belongs to an issue on our [bugtracker](https://issues.integreat-app.de/). 
You can find it there by looking for an issue with the key which is mentioned in the title of this pull request.
It starts with the keyword **WEBAPP**.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -8,43 +8,44 @@ parameters:
...
-          shouldBumpVersion: << pipeline.parameters.production >>
+          build_config: << pipeline.parameters.build_config >>
           requires:
             - build
       - deliver:
-          production: << pipeline.parameters.production >>
+          deliver_to: << pipeline.parameters.build_config >>
           requires:
             - build
             - bump_version
```
Comment (Taggotty) [inline .circleci/config.yml:46]: does that mean if you manually build integreat-debug the version number is increased?

Comment (klinzo) [reply .circleci/config.yml:46]: No it shouldn't, take a look at `bump_version`. It is only bumped if config is `integreat`. Didn't find a better solution.

File: build/webpack.config.babel.js [hunk 2]
```diff
@@ -8,29 +8,36 @@ const babelConfig = require('../.babelrc.js')
...
+    throw new Error('You need to specify a config name!')
+  } else if (!validConfigNames.includes(appConfigName)) {
+    throw new Error(`Invalid config name! Allowed configs: ${validConfigNames}`)
   }
 
-  const isDebug = env.prod === 'false'
-  const appConfigName = env.config_name || 'integreat'
-  const appConfig = require(`./${appConfigName}-config`)
+  // If prod is not specified manually, default to true for 'integreat' and 'malte' configs, false for 'integreat-debug'
+  const production = prod || ['integreat', 'malte'].includes(appConfigName)
```
Comment (Taggotty) [inline build/webpack.config.babel.js:21]: when looking at the package.json prod is not defined in Debug mode. Shouldn't it be the other way around then?

File: package.json [hunk 3]
```diff
@@ -7,10 +7,11 @@
     "npm": ">=3.8"
   },
   "scripts": {
-    "start": "webpack-dev-server --config tools/webpack.config.babel.js --progress --env.prod=false --env.config_name=integreat-debug",
-    "build": "webpack --config tools/webpack.config.babel.js --env.prod=true --env.config_name=integreat",
-    "build:debug": "webpack --config tools/webpack.config.babel.js --env.prod=false --env.config_name=integreat-debug",
-    "build:lib": "webpack --config tools/lib.webpack.config.babel.js --env.prod=true && flow-copy-source -i '**/@(__mocks__|__tests__)/**' ./src ./lib-dist && mv ./lib-dist/lib.js.flow ./lib-dist/index.js.flow",
+    "start": "webpack-dev-server --config build/webpack.config.babel.js --progress --env.config_name=integreat-debug",
+    "start:malte": "webpack-dev-server --config build/webpack.config.babel.js --progress --env.config_name=malte",
```
Comment (Taggotty) [inline package.json:11]: As stated above env.prod is not defined meaning that prod is undefined

File: build/configs/integreat/index.js [hunk 4]
```diff
@@ -1,14 +1,16 @@
 // @flow
 
-import type { AppConfigType } from '../../src/AppConfigType'
-
-import brightTheme from '../../src/modules/theme/constants/theme'
+import integreatTheme, { darkTheme as darkIntegreatTheme } from '../../themes/integreat'
+import type { AppConfigType } from '../AppConfigType'
+import featureFlags from '../featureFlags'
 
 const IntegreatAppConfig: AppConfigType = {
```
Comment (maxammann) [inline build/configs/integreat/index.js:7]: The term `AppConfig` should not be used anymore imo. As stated in the issue the term is `BuildConfig`. Each build config contains Feature Flags.

File: build/configs/integreat/index.js [hunk 5]
```diff
@@ -1,14 +1,16 @@
...
   appTitle: 'Integreat',
   itunesAppId: '1072353915',
-  theme: brightTheme,
+  theme: integreatTheme,
+  darkTheme: darkIntegreatTheme,
   cmsUrl: 'https://cms.integreat-app.de',
+  featureFlags,
```
Comment (maxammann) [inline build/configs/integreat/index.js:13]: When writing the issue I also thought that `theme` and `appTitle` etc are feature flags. So you could just add here:
```suggestion
  pois: false,
  newsStream: false
```
When talking about build configs you can say: `I set the feature flag pois is the build config malte to true`.

File: build/configs/integreat-debug/assets/apple-touch-icon.png [hunk 6]
```diff
@@ -0,0 +1 @@
+../../integreat/assets/apple-touch-icon.png
```
Comment (maxammann) [inline build/configs/integreat-debug/assets/apple-touch-icon.png:1]: Cool that symlinks work here :+1:

File: package.json [hunk 7]
```diff
@@ -7,10 +7,11 @@
...
+    "build": "webpack --config build/webpack.config.babel.js --env.config_name=integreat",
```
Comment (maxammann) [inline package.json:12]: I think it is easier to understand if we keep `--env.prod=false` here. That way determining whether you just created a production build or not is just a single "hop" away

File: .circleci/config.yml [hunk 8]
```diff
@@ -8,43 +8,44 @@ parameters:
     type: boolean
     description: Whether the pipeline was triggered through the CircleCi API (https://circleci.com/docs/api/v2/?shell#trigger-a-new-pipeline).
     default: false
-  production:
-    type: boolean
-    description: Whether to deploy to production (integreat.app) or not (webnext.integreat-app.de).
-    default: false
+  build_config:
```
Comment (maxammann) [inline .circleci/config.yml:11]: Based on my comment in the webpack config and package.json I think it makes sense to add here a boolean parameter `environment_production` or something named similar to set whether a development or production build should be made.

Comment (maxammann) [reply .circleci/config.yml:11]: That way it is possible to manually deploy:

- a malteser development build 
- integreat development build 
- integreat production build using the test cms.

File: build/configs/integreat/index.js (hunk 5 above)
Comment (klinzo) [reply build/configs/integreat/index.js:13]: As discussed: will keep this, `feature flags` is a `build option`.

File: .circleci/config.yml (hunk 8 above)
Comment (klinzo) [reply .circleci/config.yml:11]: Honestly, I don't think we want to do this, especially I'm really sure we don't want to deploy a debug build/test cms build to production.

Comment (klinzo) [reply .circleci/config.yml:11]: The only use case I'm seeing is deploying a production build to webnext for debugging purposes like the ie11 error.

Comment (klinzo) [reply .circleci/config.yml:11]: So imo we should have the following possibilites:
- Build config: Integreat, Deploy to: integreat.app, Production build: true
- Build config: Integreat test cms, Deploy to: webnext, Production build: true/false
- Build config: Malte, Deploy to malte.xyz, Production build: true
(At some point: Build config: Malte test cms, Deploy to malte webnext, Production build: true/false)
And I honestly don't now if we need the possibility to deploy a production test cms build at all.

File: .circleci/config.yml [hunk 9]
```diff
@@ -113,18 +133,9 @@ jobs:
...
-                command: yarn run build
-      - unless:
-          condition: << parameters.production >>
-          steps:
-            - run:
-                name: Debug build
-                command: yarn run build:debug
+      - run:
+          name: Build << parameters.build_config >>
+          command: yarn run webpack --config build/webpack.config.babel.js --env.config_name=<< parameters.build_config >>
```
Comment (klinzo) [inline .circleci/config.yml:138]: TODO: Specify production

File: .circleci/config.yml (hunk 8 above)
Comment (klinzo) [reply .circleci/config.yml:11]: As discussed in slack, won't do.

File: .circleci/config.yml [hunk 10]
```diff
@@ -149,30 +172,39 @@ jobs:
...
+                command: echo "Skip delivery"
 
   bump_version:
     parameters:
-      shouldBumpVersion:
-        description:
-        type: boolean
-        default: true
+      build_config:
+        description: Which build config to use. Only bump version if build_config is "integreat".
```
Comment (Taggotty) [inline .circleci/config.yml:205]: then why are "integreat-test-cms" and "malte" still part of the enum?

Comment (klinzo) [reply .circleci/config.yml:205]: We can't conditionally run jobs in circleci, therefore we have to run `bump_version` all the time we use a workflow which sometimes bumps a version and sometimes not, which is the case for the api_triggered workflow. Therefore, the passed build_config may also be malte and test cms.

Comment (klinzo) [reply .circleci/config.yml:205]: But we can conditonally run workflows... :thinking:  I guess we should do that then. Thanks!

Comment (klinzo) [reply .circleci/config.yml]: @Taggotty what do you think of my change?

File: .circleci/config.yml [hunk 11]
```diff
@@ -54,23 +71,41 @@ workflows:
...
+      - schedule:
+          cron: "0 5 1,15 * *" # For sanity checks for this syntax: https://crontab.guru/
+          filters:
+            branches:
+              only:
+                - develop
+    jobs:
+      - build:
+          build_config: "malte"
+      - deliver:
```
Comment (maxammann) [inline .circleci/config.yml:96]: This will not work as the type is `approval`. A step called like `await_production_approval` is required.

Comment (maxammann) [reply .circleci/config.yml:96]:
```
      - await_production_approval:
          type: approval
          requires:
            - bump_version
```

Comment (klinzo) [reply .circleci/config.yml:96]: No, this is not needed. Will work anyway. No extra step required.

File: .circleci/config.yml [hunk 12]
```diff
@@ -8,43 +8,60 @@ parameters:
...
             branches:
               only:
                 - develop
+          deliver_to: "integreat-test-cms"
+          requires:
+            - build
 
-  api-triggered-delivery: # Workflow for manual delivery
-    when: << pipeline.parameters.api_triggered >>
+  api-triggered-integreat-delivery: # Workflow for manual delivery of integreat which also bumps the version
```
[the remark this answers is not in the document]
Comment (klinzo) [reply .circleci/config.yml:33]: What do you mean by `commit does not deliver`?

[the remark this answers is not in the document]
Comment (maxammann) [reply .circleci/config.yml:33]: The commit workflow only delivers on develop so it makes sense that the other workflows have `delivery` in the name.

File: .circleci/config.yml (hunk 11 above)
Comment (maxammann) [reply .circleci/config.yml:96]: Did you test this? For me it did not work when I tried it. I could click approve but the job was not executed.

File: .circleci/config.yml [hunk 13]
```diff
@@ -8,43 +8,60 @@ parameters:
...
             - build
       - deliver:
-          production: << pipeline.parameters.production >>
+          deliver_to: << pipeline.parameters.build_config >>
           requires:
             - build
             - bump_version
 
-  bi-weekly-production-delivery:
+  api-triggered-delivery: # Workflow for manual delivery of integreat-test-cms and malte
```
[the remark this answers is not in the document]
Comment (klinzo) [reply .circleci/config.yml:50]: Yes, we probably should only deliver to both malte and integreat or malte-test-cms and integreat-test-cms at the same time. I'll create an issue for this. If we allow separate releases, we shouldn't bump the version for malte alone imo.

File: .circleci/config.yml (hunk 11 above)
Comment (klinzo) [reply .circleci/config.yml:96]: Yes, already approved one delivery.

Comment (klinzo) [reply .circleci/config.yml:96]: Edit: Alright, I see. Yeah seems like this is also the case for me, thanks for pointing it out.

File: .circleci/config.yml (hunk 12 above)
[the remark this answers is not in the document]
Comment (klinzo) [reply .circleci/config.yml:33]: Yeah I guess it would make sense to change the naming here.

File: .circleci/config.yml (hunk 11 above)
Comment (klinzo) [reply .circleci/config.yml:96]: @maxammann  Should I keep the required approval or should we just deliver?

File: .circleci/config.yml (hunk 13 above)
[the remark this answers is not in the document]
Comment (klinzo) [reply .circleci/config.yml:50]: I'll add this to https://issues.integreat-app.de/browse/WEBAPP-558

File: .circleci/config.yml (hunk 11 above)
Comment (maxammann) [reply .circleci/config.yml]: For malte we can remove the approval, for integreat.app we should keep it.

Comment (klinzo) [reply .circleci/config.yml]: :+1:

Comment (klinzo) [reply .circleci/config.yml:82]: @maxammann done

**Pull request merged**